lib/broadway/producer.ex

```diff
@@ -171,14 +187,64 @@ defmodule Broadway.Producer do
   end
 
   @impl true
+  def handle_info({__MODULE__, :cancel_consumers}, %{rate_limiting: %{} = rate_limiting} = state) do
+    rate_limiting = %{rate_limiting | draining?: true}
+
+    if :queue.is_empty(rate_limiting.message_buffer) do
+      cancel_consumers(state)
+    end
+
+    {:noreply, [], %{state | rate_limiting: rate_limiting}}
+  end
+
   def handle_info({__MODULE__, :cancel_consumers}, state) do
-    for from <- state.consumers do
-      send(self(), {:"$gen_producer", from, {:cancel, :shutdown}})
+    cancel_consumers(state)
+    {:noreply, [], state}
+  end
+
+  # Don't forward buffered demand when we're draining.
+  def handle_info({__MODULE__, :handle_next_demand}, %{rate_limiting: %{draining?: true}} = state) do
```

Comment:
We also don't want to forward it rate limiting is closed. So as I said in the previous comment, I would move the open/closed check to this clause too and remove the open/closed check at the moment of scheduling (i.e. we always schedule as long as the buffer is not empty).